src/app/src/main/kotlin/org/vaccineimpact/api/app/controllers/UserController.kt

```diff
@@ -14,31 +16,50 @@ class UserController(context: ControllerContext) : AbstractController(context)
             SecuredEndpoint("/", this::getUsers, setOf("*/users.read"))
     )
 
-    fun getUser(context: ActionContext): UserInterface
+    fun getUser(context: ActionContext): User
     {
         val userName = userName(context)
+        val roleReadingScopes = roleReadingScopes(context)
 
-        val roleReadingPermissions = context.permissions.filter { p -> p.name == "roles.read" }
+        val user = repos.user().use { it.getUserByUsername(userName) }
+        if (roleReadingScopes.any())
+        {
+            val allRoles = repos.user().use { it.getRolesForUser(userName) }
+            val roles = filteredRoles(allRoles, roleReadingScopes)
+            return user.copy(roles = roles)
+        }
+        else
+        {
+            return user
+        }
+    }
 
-        if (!roleReadingPermissions.any())
-            return repos.user().use { it.getUserByUsername(userName) }
+    fun getUsers(context: ActionContext): List<User>
+    {
+        val roleReadingScopes = roleReadingScopes(context)
 
-        val userWithRoles = repos.user().use { it.getUserWithRolesByUsername(userName) }
+        if (roleReadingScopes.any())
+        {
+            val users = repos.user().use { it.allWithRoles().toList() }
 
-        userWithRoles.roles = userWithRoles.roles.filter { r ->
-            roleReadingPermissions.any {
-                p ->  p.scope.encompasses(Scope.parse(r))
+            return users.map {
+                u ->
+                u.copy(roles = filteredRoles(u.roles, roleReadingScopes))
             }
         }
+        else
+        {
+            return repos.user().use { it.all().toList() }
+        }
 
-        return userWithRoles
     }
 
+    private fun userName(context: ActionContext): String = context.params(":username")
 
-    fun getUsers(context: ActionContext): List<User>
-    {
-        return return repos.user().use { it.all().toList() }
-    }
+    private fun roleReadingScopes(context: ActionContext) = context.permissions
+            .filter { it.name == "roles.read" }
+            .map { it.scope }
```

Comment:
I wonder if these private functions contain reusable logic/should be moved out into the roles class